tests/run/pyintop.pyx

```diff
@@ -515,3 +516,158 @@ def test_avoid_if_coercion(obj):
         return True
     else:
         return False
+
+@cython.test_fail_if_path_exists('//AddNode')
```

Comment:
I'd really check for a generic `BinopNode` for the tests, not a specific one. Especially not something as half-specific as `IntBinopNode`. Whether that's used or not depends on other factors than just this PR. We're not expecting any `BinopNode` at all in the test functions and don't want the tests to become no-ops by accident when we add other optimisations.

Reply:
> I'd really check for a generic `BinopNode` for the tests, not a specific one. 

When trying to substitude `'//AddNode'` with `'//BinopNode'` the test is not failing when the fix is not present. I suppose tests added should fail when the fix is not present (e.g. on master branch). Moreover, when I print the tree of the test there is no `BinopNode` present:
```
          - stats[0]: DefNode(name="pure_add_int")
            - args[0]: CArgDeclNode(pos=(pyintop.pyx:522:17))
              - base_type: CAnalysedBaseTypeNode(pos=(pyintop.pyx:522:23))
              - declarator: CNameDeclaratorNode(pos=(pyintop.pyx:522:17))
              - annotation: AnnotationNode(type=None)
            - body: StatListNode(pos=(pyintop.pyx:523:4))
              - stats[0]: CompilerDirectivesNode(pos=(pyintop.pyx:523:4))
                - body: StatListNode(pos=(pyintop.pyx:523:4))
                  - stats[0]: SingleAssignmentNode(pos=(pyintop.pyx:527:16))
                    - lhs: NameNode(type=<PyObjectType>, name="res1")
                    - rhs: AddNode(type=<PyObjectType>)
                      - operand1: NameNode(type=<PyInt_Type>, name="obj2")
                      - operand2: IntNode(type=<PyObjectType>, value='1')
```
That's why I used specific nodes instead of generic `BinopNode`. Am I missing something?

Reply:
Don't think so. At least according to the wiki (https://github.com/cython/cython/wiki/HackerGuide#parse-tree-assertions)

> You can express assertions using a simple XPath-like language called TreePath that traverses the parse tree. Nodes are referred to by their type name (inheritance is not considered).

It's possible we might want to improve that but I don't think as part of this PR.

-----------------------

One thing you could do is add separate "fail" tests for the base classes. E.g. this one would be 

```
@cython.test_fail_if_path_exists('//AddNode')
@cython.test_fail_if_path_exists('//NumBinopNode')
@cython.test_fail_if_path_exists('//BinopNode')
```

That gives a higher chance of catching it if some optimization changes

Reply:
Fixed as proposed by @da-woods .